eternalcombat-plugin/src/main/java/com/eternalcode/combat/border/particle/BorderBlockController.java

```diff
@@ -0,0 +1,207 @@
+
+    public void playBlocks(Player player, Collection<BorderPoint> blocks) {
+        Map<Vector3i, Set<BorderPoint>> chunksToUpdate = this.splitToChunks(blocks);
+
+        for (Map.Entry<Vector3i, Set<BorderPoint>> entry : chunksToUpdate.entrySet()) {
+            Vector3i chunk = entry.getKey();
+            EncodedBlock[] encodedBlocks = entry.getValue().stream()
+                .map(borderPoint -> this.toEncodedBlock(borderPoint))
+                .toArray(value -> new EncodedBlock[value]);
+
+            WrapperPlayServerMultiBlockChange multiBlockChange = new WrapperPlayServerMultiBlockChange(chunk,
+                true,
+                encodedBlocks
+            );
+
+            PLAYER_MANAGER.sendPacket(player, multiBlockChange);
+        }
+    }
+
+    public void playHideBlocks(Player player, Collection<BorderPoint> blocks) {
+        Map<Vector3i, Set<BorderPoint>> chunksToUpdate = splitToChunks(blocks);
+
+        for (Map.Entry<Vector3i, Set<BorderPoint>> entry : chunksToUpdate.entrySet()) {
+            Vector3i chunk = entry.getKey();
+            EncodedBlock[] encodedBlocks = entry.getValue().stream()
+                .map(point -> new EncodedBlock(AIR_ID, point.x(), point.y(), point.z()))
+                .toArray(value -> new EncodedBlock[value]);
+
+            WrapperPlayServerMultiBlockChange multiBlockChange = new WrapperPlayServerMultiBlockChange(chunk,
+                true,
+                encodedBlocks
+            );
+
+            PLAYER_MANAGER.sendPacket(player, multiBlockChange);
+        }
+    }
+
+    private Set<BorderPoint> processChunks(Player player, Collection<BorderPoint> blocks) {
+        Map<ChunkLocation, Set<BorderPoint>> chunksToProcess = new HashMap<>();
+        for (BorderPoint block : blocks) {
+            BorderPoint inclusive = block.toInclusive();
+            ChunkLocation location = new ChunkLocation(inclusive.x() >> 4, inclusive.z() >> 4);
+            chunksToProcess.computeIfAbsent(location, k -> new HashSet<>()).add(inclusive);
+        }
+
+        Set<BorderPoint> points = new HashSet<>();
+        chunksToProcess.forEach((location, chunkPoints) -> {
+            ChunkSnapshot snapshot = loadChunkSnapshot(player, location);
+            if (snapshot == null) {
+                return;
+            }
+
+            for (BorderPoint point : chunkPoints) {
+                Material type = snapshot.getBlockType(point.x() - (location.x() << 4), point.y(), point.z() - (location.z() << 4));
+                if (!type.isAir()) {
+                    continue;
+                }
+
+                points.add(point);
+            }
+        });
+
+        return points;
+    }
+
+    private Map<Vector3i, Set<BorderPoint>> splitToChunks(Collection<BorderPoint> blocks) {
+        Map<Vector3i, Set<BorderPoint>> chunksToUpdate = new HashMap<>();
+        for (BorderPoint block : blocks) {
+            Vector3i chunk = new Vector3i(
+                block.x() >> 4,
```

Comment:
You can also extract it to a constant field like you did with the CHUNK_SHIFT before. I have no idea what this `4` is supposed to mean alone